File: src/DashSpec.Core/Parsing/CardChromeParser.cs
using DashSpec.Core.Model;

namespace DashSpec.Core.Parsing;

internal static class CardChromeParser
{
    public static CardChromeDefinition Parse(TokenReader reader, string cardId)
    {
        BlockSyntax.BeginBlock(reader);
        CardBoundFilterChrome boundFilters = CardBoundFilterChrome.Chips;

        while (!BlockSyntax.IsBlockEnd(reader, "chrome"))
        {
            reader.SkipNewlines();
            if (BlockSyntax.IsBlockEnd(reader, "chrome"))
            {
                break;
            }

            if (reader.TryKeyword("bound_filters"))
            {
                reader.Expect(TokenKind.Eq);
                boundFilters = reader.ReadIdent().Trim().ToLowerInvariant() switch
                {
                    "hidden" => CardBoundFilterChrome.Hidden,
                    "toolbar_only" or "toolbar-only" or "toolbaronly" => CardBoundFilterChrome.ToolbarOnly,
                    "chips" => CardBoundFilterChrome.Chips,
                    var raw => throw new DashSpecParseException(
                        $"Card '{cardId}': chrome bound_filters must be chips, hidden, or toolbar_only; got '{raw}'."),
                };

                reader.SkipNewlines();
                continue;
            }

            throw reader.Unexpected("chrome property");
        }

        BlockSyntax.ExpectBlockEnd(reader, "chrome");
        return new CardChromeDefinition(boundFilters);
    }
}

internal static class FilterDeriveParser
{
    public static FilterDeriveDefinition Parse(TokenReader reader, string pageId)
    {
        var target = reader.ReadIdent();
        if (!reader.TryKeyword("from"))
        {
            throw new DashSpecParseException($"Page '{pageId}': derive requires 'from <filter>'.");
        }

        var source = reader.ReadIdent();
        string? grainFilter = null;
        if (reader.TryKeyword("grain"))
        {
            grainFilter = reader.ReadIdent();
        }

        reader.SkipNewlines();
        return new FilterDeriveDefinition(target, source, grainFilter);
    }
}

internal static class ToolbarBoardFactory
{
    public static LayoutBoardDefinition FromFilterNames(IReadOnlyList<string> names)
    {
        if (names.Count == 0)
        {
            throw new DashSpecParseException("toolbar requires at least one filter name.");
        }

        return new LayoutBoardDefinition([names.ToList()]);
    }
}

File: src/DashSpec.Modeling.Parse/Card/CardChromeParser.fs
namespace DashSpec.Modeling.Parse.Card

open System
open System.Collections.Generic
open DashSpec.Modeling.Core
open DashSpec.Modeling.Parse
open DashSpec.Modeling.Parse.Layout
open DashSpec.Modeling.Parse.Lexing

module CardChromeParser =

    let parse (reader: TokenReader) (cardId: string) =
        BlockSyntax.beginBlock reader
        let mutable boundFilters = CardBoundFilterChrome.Chips

        while not (BlockSyntax.isBlockEnd reader "chrome" None) do
            reader.SkipNewlines()
            if BlockSyntax.isBlockEnd reader "chrome" None then ()
            elif reader.TryKeyword "bound_filters" then
                reader.Expect TokenKind.Eq
                boundFilters <-
                    match reader.ReadIdent().Trim().ToLowerInvariant() with
                    | "hidden" -> CardBoundFilterChrome.Hidden
                    | "toolbar_only" | "toolbar-only" | "toolbaronly" -> CardBoundFilterChrome.ToolbarOnly
                    | "chips" -> CardBoundFilterChrome.Chips
                    | raw ->
                        raise (DashSpecParseException($"Card '{cardId}': chrome bound_filters must be chips, hidden, or toolbar_only; got '{raw}'."))
                reader.SkipNewlines()
            else
                raise (reader.Unexpected "chrome property")

        BlockSyntax.expectBlockEnd reader "chrome" None
        { BoundFilters = boundFilters }

module FilterDeriveParser =

    let parse (reader: TokenReader) (pageId: string) =
        let target = reader.ReadIdent()
        if not (reader.TryKeyword "from") then
            raise (DashSpecParseException($"Page '{pageId}': derive requires 'from <filter>'."))
        let source = reader.ReadIdent()
        let grainFilter =
            if reader.TryKeyword "grain" then Some(reader.ReadIdent())
            else None
        reader.SkipNewlines()
        { TargetFilter = target; SourceFilter = source; GrainFilterName = grainFilter }

module ToolbarBoardFactory =

    let fromFilterNames (names: IReadOnlyList<string>) =
        if names.Count = 0 then
            raise (DashSpecParseException("toolbar requires at least one filter name."))
        { Rows = [| names |] :> IReadOnlyList<_>; ModuleScope = None }
